File: flutter_segment/android/src/main/kotlin/com/claimsforce/flutter_segment/Segment.kt
package com.claimsforce.flutter_segment

import android.content.Context
import com.segment.analytics.Analytics
import com.segment.analytics.Options
import com.segment.analytics.Properties
import com.segment.analytics.Traits
import com.segment.analytics.integrations.Integration

class Segment(private val applicationContext: Context) {
    private val integrations: Array<Integration.Factory> = emptyArray()

    fun configure(writeKey: String, trackApplicationLifecycleEvents: Boolean, debug: Boolean) {
        val analyticsBuilder = Analytics.Builder(applicationContext, writeKey)

        if (trackApplicationLifecycleEvents) {
            analyticsBuilder.trackApplicationLifecycleEvents()
        }
        if (debug) {
            analyticsBuilder.logLevel(Analytics.LogLevel.DEBUG)
        }

        for (integration in integrations) {
            analyticsBuilder.use(integration)
        }

        Analytics.setSingletonInstance(analyticsBuilder.build())
    }
    
    fun identify(userId: String, traits: Traits, options: Options) {
        Analytics.with(this.applicationContext).identify(userId, traits, options)
    }

    fun track(eventName: String, properties: Properties, options: Options) {
        Analytics.with(this.applicationContext).track(eventName, properties, options)
    }

    fun screen(screenName: String, properties: Properties, options: Options) {
        Analytics.with(this.applicationContext).screen(null, screenName, properties, options)
    }

    fun group(groupId: String, traits: Traits, options: Options) {
        Analytics.with(this.applicationContext).group(groupId, traits, options)
    }

    fun alias(alias: String, options: Options) {
        Analytics.with(this.applicationContext).alias(alias, options)
    }

    fun getAnonymousId(): String {
        return Analytics.with(this.applicationContext).analyticsContext.traits().anonymousId()
    }

    fun reset() {
        Analytics.with(this.applicationContext).reset()
    }

    fun disable() {
        Analytics.with(this.applicationContext).optOut(true)
    }

    fun enable() {
        Analytics.with(this.applicationContext).optOut(false)
    }

    fun setContext(context: HashMap<String, Any>) {
        // TODO
    }
}

File: flutter_segment/lib/src/segment.dart
import 'dart:async';

import 'package:flutter_segment_platform_interface/flutter_segment_platform_interface.dart';

class Segment {
  static FlutterSegmentPlatform get _platform => FlutterSegmentPlatform.instance;

  static Future<void> configure(String writeKey, {
    bool trackApplicationLifecycleEvents = false,
    bool debug = false,
  }) {
    return _platform.configure(
      writeKey,
      trackApplicationLifecycleEvents: trackApplicationLifecycleEvents,
      debug: debug,
    );
  }

  static Future<void> identify(String userId, {
    Map<String, dynamic> traits,
    Map<String, dynamic> options,
  }) {
    return _platform.identify(
        userId,
        traits: traits,
        options: options
    );
  }

  static Future<void> track(String eventName, {
    Map<String, dynamic> properties,
    Map<String, dynamic> options,
  }) {
    return _platform.track(
        eventName,
        properties: properties,
        options: options
    );
  }

  static Future<void> screen(String screenName, {
    Map<String, dynamic> properties,
    Map<String, dynamic> options,
  }) {
    return _platform.screen(
        screenName,
        properties: properties,
        options: options
    );
  }

  static Future<void> group(String groupId, {
    Map<String, dynamic> traits,
    Map<String, dynamic> options,
  }) {
    return _platform.group(
        groupId,
        traits: traits,
        options: options
    );
  }

  static Future<void> alias(String alias, {
    Map<String, dynamic> options,
  }) {
    return _platform.alias(
        alias,
        options: options
    );
  }

  static Future<String> get getAnonymousId {
    return _platform.getAnonymousId;
  }

  static Future<void> reset() {
    return _platform.reset();
  }

  static Future<void> disable() {
    return _platform.disable();
  }

  static Future<void> enable() {
    return _platform.enable();
  }

  static Future<void> debug(bool enabled) {
    return _platform.debug(enabled);
  }

  static Future<void> setContext(Map<String, dynamic> context) {
    return _platform.setContext(context);
  }
}
